Qml modules

chubinou opened the pull request:
This MR aims to bring the support of QML modules to meson, the goal is to
provide something similar to CMake `qt_add_qml_module` function provided by Qt
(see https://doc.qt.io/qt-6/qt-add-qml-module.html)

fixes: #6988 #9683

what's provided

* qml module generation, qmldir and qrc are automatically generated
* automatic type registration, C++ classes can be automatically registered in
  the QML module
* cache generation, the qmlcachegen tool can be invoked on qml files.

what's missing:

the scope of this PR is already large, I voluntary put aside some aspect of the module generation.

* no qmltc integration
* no plugin code generation
* no linter target
* no qml import scanning
* no failsafe regarding which tool is available with a given version of Qt, some
  tools (or options) are not available with older version of Qt,
* no 'past version' support, this would require being able to tag version to
  specific Qml files, which is not very practical in terms of API
* cmake copies every qml in the build directory, maybe that's necessary for qml
  tooling, I haven't investigated this

| CMake argument              | support | comment |
|-----------------------------|---------|---------|
| `VERSION`                   | yes |   |
| `PAST_MAJOR_VERSIONS`       | no  | shipping different version of the same unit would probably require to be able to tag version per file |
| `STATIC / SHARED`           | N/A | we don't generated the final target, the output of the function should be added to a lib/exe target |
| `PLUGIN_TARGET`             | no  | plugin is not generated |
| `OUTPUT_DIRECTORY`          | yes |   |
| `RESOURCE_PREFIX`           | yes |   |
| `CLASS_NAME`                | no  | plugin is not generated |
| `TYPEINFO`                  | yes |   |
| `IMPORTS`                   | yes |   |
| `OPTIONAL_IMPORTS`          | yes |   |
| `DEFAULT_IMPORTS`           | yes |   |
| `DEPENDENCIES`              | yes |   |
| `IMPORT_PATH `              | yes |   |
| `SOURCES `                  | yes | source can be provided for moc, sources should be added to the final target |
| `QML_FILES `                | yes |   |
| `RESOURCES`                 | no  | resources can be added to the final target independently, there is no point supporting it there |
| `OUTPUT_TARGETS`            | N/A |   |
| `DESIGNER_SUPPORTED`        | yes |   |
| `FOLLOW_FOREIGN_VERSIONING` | no  |   |
| `NAMESPACE`                 | yes |   |
| `NO_PLUGIN`                 | no  | plugin is not generated |
| `NO_PLUGIN_OPTIONAL`        | no  | plugin is not generated |
| `NO_CREATE_PLUGIN_TARGET`   | no  | plugin is not generated |
| `NO_GENERATE_PLUGIN_SOURCE` | no  | plugin is not generated |
| `NO_GENERATE_QMLTYPES`      | yes |   |
| `NO_GENERATE_QMLDIR`        | yes |   |
| `NO_LINT`                   | no  | qmllint not supported  |
| `NO_CACHEGEN`               | yes |   |
| `NO_RESOURCE_TARGET_PATH`   |     |   |
| `NO_IMPORT_SCAN`            | no  | not supported |
| `ENABLE_TYPE_COMPILER`      | no  | qmltc is not supported |
| `TYPE_COMPILER_NAMESPACE`   | no  | qmltc is not supported |
| `QMLTC_EXPORT_DIRECTIVE`    | no  | qmltc is not supported |
| `QMLTC_EXPORT_FILE_NAME`    | no  | qmltc is not supported |

chubinou commented:
arch tests are failing because  the CI doesn't provide QtQml for qt 6 which is a new requirement for the test

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -91,10 +95,101 @@ class CompileTranslationsKwArgs(TypedDict):
+def _list_in_set_validator(choices: T.Set[str]) -> T.Callable[[T.List[str]], T.Optional[str]]:
+    """Check that the choice given was one of the given set."""
+    def inner(checklist: T.List[str]) -> T.Optional[str]:
+        for check in checklist:
+            if check not in choices:
```
I think you can simplify (and make a better error message) this to something like:
```python
invalid = set(checklist).difference(choices)
if invalid:
    return f"invalid selections {', '.join(sorted(invalid))}, valid elements are {', '.join(sorted(choices))}."

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -75,11 +77,13 @@ class PreprocessKwArgs(TypedDict):
     class HasToolKwArgs(kwargs.ExtractRequired):
 
         method: str
+        tools: T.List[str]
```
you can make this a little more powerful (and maybe remove an assert later on) with:
`tools: T.List[Literal['moc, 'lrelease', <insert reset of valid tools>]]`

chubinou replied:
fixed

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -616,3 +734,364 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+        elif isinstance(source, (build.CustomTarget, build.CustomTargetIndex)):
+            output: str = source.get_filename()
+            return os.path.join(state.environment.get_build_dir(), state.backend.get_target_dir(source), output)
+        else:
+            assert False, "unreachable"
```
I'm happy to trust the type annotations for the final release, though this might be useful for bringup

chubinou replied:
I changed this function quite a bit, I took inspiration from the one in gnome.py

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -616,3 +734,364 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+                    basename: str = os.path.basename(sourcepath)
+                    classname: str = basename.rsplit('.', maxsplit=1)[0]
+
+                    if not any(basename.endswith(ext) for ext in ['.qml', '.js', '.mjs']):
```
```suggestion
                    if not basename.endswith(('.qml', '.js', '.mjs')):
```

chubinou replied:
fixed

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -616,3 +734,364 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+                    if not any(basename.endswith(ext) for ext in ['.qml', '.js', '.mjs']):
+                        raise MesonException(f'unexpected file type declared in qml sources {s}')
+
+                    if len(classname) == 0 or '.' in classname or classname[0].islower():
```
```suggestion
                    if not classname or '.' in classname or classname[0].islower():
```

chubinou replied:
fixed

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -616,3 +734,364 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+
+        cmd.extend(kwargs['extra_args'])
+
+        if namespace != '':
```
```suggestion
        if namespace:
```

chubinou replied:
fixed

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -616,3 +734,364 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+            description=f'Qml type registration for {target_name}',
+        )
+
+    @FeatureNew('qt.qml_module', '1.0')
```
```suggestion
    @FeatureNew('qt.qml_module', '1.6')
```

chubinou replied:
fixed

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -616,3 +734,364 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+        os.makedirs(module_path_outdir, exist_ok=True)
+
+        qml_qresource_abs: str = os.path.join(module_path_outdir, 'qml.qrc')
+        qrc_resouces: T.List[T.Union[FileOrString, build.CustomTarget, build.CustomTargetIndex, build.GeneratedList]] = []
+        all_qml_files: T.Sequence[T.Union[FileOrString, build.CustomTarget]] = list(kwargs['qml_sources']) + list(kwargs['qml_singletons']) + list(kwargs['qml_internals'])
```
I'm wondering if it would be beneficial to use the `Interpreter.source_strings_to_files` method here to remove the strings and just have `list[File | CustomTarget | CustomTargetIndex | GeneratedList]` here, and then not have to do string handling internally.

dcbaker reviewed (changes requested):
Thanks for working on this! I've taken a cursory glance, and there's a few smallish things I've noticed. I'd like to have another look later when I have more time to get into more of the details

chubinou commented:
some self comments on the PR:

I tried to follow Qt module directory structure

```
for the file MyClass.qml in the module Foo.Bar, this should looks like
Foo
|- Bar
|   |- qmldir
|   |- MyClass.qml
|   |- module.qmltypes
```

at the moment this is still a bit clumpsy:

* for the tools to actually be able to use this, we should copy the .qml files to the Bar directory, but Meson doesn't really let me copy files to a subdirectory (I can probably cheat my way out of it, but choosed not to for the moment)
* this is not really needed for the compilation parts, we could generate everything on a flat level, and do the file mapping in the qrc (it defines the internal paths of each resources).  I thought that maybe I can make a special target to install the qml module tree for tooling purpose, maybe some custom target with an `install_tag` like in the gnome module, I haven't looked in depth at this.

AdelKS commented:
Thank you for undertaking this work !

I wanted to give a try to registering QML modules but I am afraid I've hit against a regression introduced in this MR before I could get to it. Here's the reproducer:
```shell
git clone --depth=1 https://github.com/AdelKS/ZeGrapher.git
cd ZeGrapher
meson setup build
cd build
meson compile
```
which outputs this error
```
[59/203] Compiling C++ object src/libZeGrapherLib.a.p/meson-generated_moc_polynomialregression.cpp.o
FAILED: src/libZeGrapherLib.a.p/meson-generated_moc_polynomialregression.cpp.o 
c++ -Isrc/libZeGrapherLib.a.p -Isrc -I../src -Isubprojects/zecalculator/include -I../subprojects/zecalculator/include -I/usr/include/qt6/QtCore -I/usr/include/qt6 -I/usr/lib/qt6/mkspecs/linux-g++ -I/usr/include/qt6/QtGui -I/usr/include/qt6/QtWidgets -I/usr/include/qt6/QtNetwork -I/usr/include/qt6/QtSvg -I/usr/include/qt6/QtQuick -I/usr/include/qt6/QtQmlModels -I/usr/include/qt6/QtQmlIntegration -I/usr/include/qt6/QtOpenGL -I/usr/include/qt6/QtQml -I/usr/include/qt6/QtQuickWidgets -I/usr/include -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -std=c++20 -O3 '-D QT_NO_DEBUG_OUTPUT' '-D QT_NO_INFO_OUTPUT' '-D QT_NO_WARNING_OUTPUT' '-D QT_NO_DEBUG' -pipe -fPIC -DBOOST_ALL_NO_LIB -DQT_QUICKWIDGETS_LIB -DQT_QUICK_LIB -DQT_QMLMODELS_LIB -DQT_QMLINTEGRATION_LIB -DQT_OPENGL_LIB -DQT_QML_LIB -DQT_SVG_LIB -DQT_NETWORK_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_CORE_LIB -MD -MQ src/libZeGrapherLib.a.p/meson-generated_moc_polynomialregression.cpp.o -MF src/libZeGrapherLib.a.p/meson-generated_moc_polynomialregression.cpp.o.d -o src/libZeGrapherLib.a.p/meson-generated_moc_polynomialregression.cpp.o -c src/libZeGrapherLib.a.p/moc_polynomialregression.cpp
cc1plus: fatal error: src/libZeGrapherLib.a.p/moc_polynomialregression.cpp: No such file or directory
```
it looks like there's a naming convention that isn't respected between `meson-generated_moc_polynomialregression.cpp.o ` and `moc_polynomialregression.cpp` ? The relevant `meson` file [is this one](https://github.com/AdelKS/ZeGrapher/blob/master/src/meson.build)

chubinou commented:
Hi, thanks for your interest 

> I wanted to give a try to registering QML modules but I am afraid I've hit against a regression introduced in this MR before I could get to it. 

your repository doesn't use the feature from this MR, so I don't really see how this is a regression.

I tried to compile you project but I ended up with a lot of compilation issues (template issues in your math classes, compiled on ubuntu 22.04 with GCC 11.4), So I can't really help you there

> Or maybe it's related to the fact that I currently pass all headers that need a `moc` pass on them to `meson`'s `qt.compile_moc()` method. I wanted to ask how does that interact with the new `qt.qml_module()` method with its `moc_headers` kwarg, because it looked like `ninja` was already complaining that a target is generated twice if I passed a header, that I already `moc`'ed with `qt.compile_moc()` , to `qt.qml_module()`

qt.qml_module will moc the file itself as it needs to pass specific options to moc to extract type information, so `qt.compile_moc` becomes redundant

AdelKS commented:
> I need to try latest meson git commit to check, maybe my build failure has nothing to do with this MR.

Okay I just tried on meson commit `dfd22db4b`, which is the latest meson commit before your MRs commits, and my repo could build, so this MR did at least introduce breaking changes, to `qt.compile_moc()` I assume ?

Would you be willing to install `gcc-13` on your Ubuntu to confirm ?
```shell
add-apt-repository -y ppa:ubuntu-toolchain-r/test
apt install -y gcc-13
git clone --depth=1 https://github.com/AdelKS/ZeGrapher.git
cd ZeGrapher
CXX=g++13 meson setup build
cd build
meson compile
```

ferdnyc commented on mesonbuild/modules/_qt.py:
```diff
@@ -269,8 +284,9 @@ def has_tools(self, state: 'ModuleState', args: T.Tuple, kwargs: 'HasToolKwArgs'
-        for tool in self.tools.values():
-            if not tool.found():
+        for tool in kwargs['tools']:
+            assert tool in self.tools, 'tools must be in {moc, uic, rcc, lrelease}'
```
This is the third time that same set of tools (`moc, uic, rcc, lrelease`) is explicitly listed out in the code, hardcoded — that's a lot of redundancy and places to keep in sync if the list ever changes. Couldn't it be defined somewhere as a set variable, then used...

* on line 273:

```python
  default=list(_set_of_qt_tools)
```
* on line 274:
 
```python
QtBaseModule._list_in_set_validator(_set_of_qt_tools)
```

* and here:
 
```python
assert tool in self.tools, f'tools must be in {_set_of_qt_tools}'
```

to keep things from potentially getting out of sync?

(Or, heck, at least here as...)

```python
assert tool in self.tools, f'tools must be in {self.tools}'
```

chubinou replied:
fixed, though it doesn't seems that I can reference _set_of_qt_tools for the typechecker in `HasToolKwArgs`

ferdnyc replied:
Mm, it seems the arg is typed as a `T.List[str]`... perhaps that should be more generic.

Hmm. _Is_ there a generic that covers both `list` and `set`? Neither `Sequence` nor `Mapping` will match both. (I suppose you could call `HasToolKwArgs(list(_set_of_qt_tools))` but that's kind of ugh.)

chubinou replied:
patch has been moved to #13632 (see discussions in #13524). 

> Hmm. Is there a generic that covers both list and set

I don't know about this

eli-schwartz replied:
resolving conversation -- the patch is merged there and should be dropped from this PR in the next rebase.

chubinou commented:
@AdelKS the issue was due to some bad replacements in ninja backend, i opened #13526 to fix the issue

chubinou commented:
I updated the PR, it no longer requires the sources to be generated with a precise directory structure which should match meson philosophy.

I also added a 'devel' install target to allow installing the module definition (for qmllint or and qml tools)

jpakkane commented on mesonbuild/modules/_qt.py:
```diff
@@ -616,3 +733,431 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+        KwargInfo('method', str, default='auto'),
+        KwargInfo('preserve_paths', bool, default=False),
+    )
+    def qml_module(self, state: ModuleState, args: T.Tuple[str, str, str], kwargs: 'QmlModuleKwArgs') -> ModuleReturnValue:
```
The type of `args` is not correct. Nothing in Meson prevents people from calling it with more or fewer arguments. It should be defined as a list of strings. It should be a `T.List[str]`. The code below should also check that its size is exactly 3 and that every entry is actually a string and error out if they are not.

eli-schwartz replied:
Well, normally we'd do that with typed_pos_args, surely. :)

chubinou replied:
I went for `typed_pos_args`

eli-schwartz commented on mesonbuild/modules/_qt.py:
```diff
@@ -477,20 +482,28 @@ def _compile_moc_impl(self, state: ModuleState, kwargs: MocCompilerKwArgs) -> T.
 
         output: T.List[build.GeneratedList] = []
 
+        do_output_json : bool = self._moc_supports_json and kwargs['output_json']
```
We always use depfiles and always `mlog.warning()` if they are not supported. For output_json, it's a kwarg so we should error out if it is not supported rather than merely skipping it.

chubinou replied:
I raised an exception here

eli-schwartz commented on mesonbuild/modules/_qt.py:
```diff
@@ -648,3 +733,431 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+            raise MesonException('qt.qml_module is not suppported for this version of Qt')
+
+        #Major.Minor(.Patch)
+        version_re = re.compile('^(\\d+)\\.(\\d+)(\\.(\\d+))?$')
```
You can simplify this a bit by using `r'string'` (officially r for raw, not r for regex, but we all know the truth, eh?)

chubinou replied:
updated,

note I moved the module regexp definition outside the class as it was a bit convoluted and reused multiple times

eli-schwartz commented on mesonbuild/modules/_qt.py:
```diff
@@ -648,3 +733,431 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+        module_prefix_full: str = os.path.join(*(kwargs['resources_prefix'].split('/') + module_prefix_list))
+
+        qrc_resouces: T.List[T.Union[FileOrString, build.GeneratedTypes]] = []
+        all_qml: T.Sequence[T.Union[FileOrString, build.GeneratedTypes]] = list(kwargs['qml_sources']) + list(kwargs['qml_singletons']) + list(kwargs['qml_internals'])
```
Aren't these already lists? Why cast them again to list?

chubinou replied:
mypy fails otherwise

```
Unsupported left operand type for + ("Sequence[File | str | CustomTarget | CustomTargetIndex | GeneratedList]")
```

eli-schwartz replied:
Maybe the original definition of kwargs should annotate these as T.List instead of the generic T.Sequence, if we need them to be a List type?

chubinou replied:
I changed it for a T.List, this seems to be enough

eli-schwartz commented on mesonbuild/modules/_qt.py:
```diff
@@ -648,3 +733,431 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+        all_qml_files: T.List[File] = self._source_to_files(state, all_qml)
+        all_qml_basename: T.List[str] = [os.path.basename(p.fname) for p in all_qml_files]
+
+        install_dir: str = kwargs['install_dir'] or "qml"
```
Nit: inconsistent use of double quotes, not the only place with this inconsistency.

chubinou replied:
fixed, 

I only updated new code, as I'm not sure  what's the prefered policy regading this (use `"` if string contains a `'`, `"""` or `'''` for class/function documentation

eli-schwartz commented on mesonbuild/modules/_qt.py:
```diff
@@ -648,3 +733,431 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+
+        #copy QML files for Qt tools
+        if kwargs['install']:
+            self.interpreter.install_data_impl(
+                all_qml_files,
+                module_install_dir,
+                FileMode(),
+                all_qml_basename,
+                'devel')
```
This looks very cramped...

chubinou replied:
updated

eli-schwartz commented on mesonbuild/modules/_qt.py:
```diff
@@ -648,3 +733,431 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+        for s in sources:
+            if isinstance(s, (build.CustomTarget, build.CustomTargetIndex)):
+                for o in s.get_outputs():
+                    content_files.append(File.from_built_file(
+                        state.backend.get_target_dir(s),
+                        o
+                    ))
```
This can be one line...

chubinou replied:
changed

eli-schwartz commented on mesonbuild/modules/_qt.py:
```diff
@@ -648,3 +733,431 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
         else:
             return ModuleReturnValue(translations, [translations])
+
+    def _source_to_files(self, state: 'ModuleState', sources: T.Sequence[T.Union[FileOrString, build.GeneratedTypes]]) -> T.List[File]:
```
Quoted annotations are no longer necessary since we consistently use `from __future__ import annotations` in all files.

chubinou replied:
fixed, I also added a commit to update align the rest of the class on this policy

eli-schwartz replied:
Sure, I guess.

Both styles "work" but the newer style allows for shorter lines (by two columns per annotation). In general we haven't bothered to fix old code, but it's always best to have new code so things the preferred way. Some old code has been naturally rewritten over time for unrelated reasons, and inherited the new annotation style along the way, so we will eventually converge on a consistent codebase.

Fixing some of them immediately is fine (though I should rebase and update my PR that adds an ignore-revs file for `git blame`!) especially if you're going to be working on this file a lot and would find it helpful. :D

I just wouldn't ever demand to fix the whole file in a review request.

jpakkane commented:
The next release is creeping closer so it would be nice to get this in. FWICT there are only some minor fixes + a rebase and this should be good to merge.

chubinou commented:
Sorry for the late update, I was first waiting for #13526 to get merged then I somehow missed the review notification late October

I made a few updates:

* I removed the need to pass a target name, I instead derive a name from the module name. The derived name matches what is done by qmltyperegistar.
* Module version is made optional too, it will default to 254.254 like cmake `qt_add_qml_module`
* module can be installed with the `devel` tag, this allows qt tools (linter, designer) to use the module definition. Though, this makes the test suite fail if Qt6 is not found as the `test.json` expects these files to be installed,  I'm not sure how to fix this.
* added qt6-declarative dependency for arch CI
* I added documentation for the function
*  code is rebased on master

dcbaker commented on mesonbuild/modules/_qt.py:
```diff
@@ -103,13 +188,23 @@ def inner(checklist: T.List[str]) -> T.Optional[str]:
     _set_of_qt_tools = {'moc', 'uic', 'rcc', 'lrelease'}
+    _moc_supports_json = False
+    _support_qml_module = False
+    _set_of_qt_tools = {'moc', 'uic', 'rcc', 'lrelease', 'qmlcachegen', 'qmltyperegistrar'}
```
The CodeQL warning about this being defined twice is true.

eli-schwartz commented:
> * module can be installed with the `devel` tag, this allows qt tools (linter, designer) to use the module definition. Though, this makes the test suite fail if Qt6 is not found as the `test.json` expects these files to be installed,  I'm not sure how to fix this.

This might only be possible by adding a new frameworks test specifically to handle qml, which can raise a SKIP if qt6 isn't installed. As opposed to the current test which simply tests all currently-installed Qt versions.

> * added qt6-declarative dependency for arch CI

Right, so that should mean the CI failure for that specific runner can be ignored. (The CI image builder for Arch does succeed, which is all we need.)

AdelKS commented:
Hey, thanks for #13526 and for this MR. 

I tried to move one of the modules of my projects to using this [small change](https://github.com/AdelKS/ZeGrapher/commit/60d5163a6533f9bbc9d3b080cb37859a73ed051e) and I've hit against some issues
- `qmlcachegen` complains about wrong input
- The generated `.cpp` file that calls `qmlRegisterTypesAndRevisions` seems to not include the header I've given

<details>
<summary>Error output</summary>

```
[1/45] Qml cache loader for zegrapher_highlighter
Usage: /usr/lib64/qt6/libexec/qmlcachegen [options] [qml file]

Options:
  -h, --help                                   Displays help on commandline
                                               options.
  --help-all                                   Displays help, including generic
                                               Qt options.
  -v, --version                                Displays version information.
  --bare                                       Do not include default import
                                               directories. This may be used to
                                               run qmlcachegen on a project
                                               using a different Qt version.
  --filter-resource-file                       Filter out QML/JS files from a
                                               resource file that can be cached
                                               ahead of time instead
  --resource-file-mapping <old-name=new-name>  Path from original resource file
                                               to new one
  --resource <resource-file-name>              Qt resource file that might
                                               later contain one of the compiled
                                               files
  --resource-path <resource-path>              Qt resource file path
                                               corresponding to the file being
                                               compiled
  --resource-name <compiled-file-list>         Required to generate
                                               qmlcache_loader without qrc
                                               files. This is the name of the Qt
                                               resource the input files belong
                                               to.
  -i <qmldir file>                             Import extra qmldir
  -I <import directory>                        Look for QML modules in
                                               specified directory
  --only-bytecode                              Generate only byte code for
                                               bindings and functions, no C++
                                               code
  --verbose                                    Output compile warnings
  --warnings-are-errors                        Treat warnings as errors
  --validate-basic-blocks                      Performs checks on the basic
                                               blocks of a function compiled
                                               ahead of time to validate its
                                               structure and coherence
  --dump-aot-stats                             Dumps statistics about
                                               ahead-of-time compilation of
                                               bindings and functions
  --module-id <id>                             Identifies the module of the qml
                                               file being compiled for aot stats
  -o <file name>                               Output file name

Arguments:
  [qml file]                                   QML source file to generate
                                               cache for.
[2/45] Compiling C++ object src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmlcache_loader.cpp.o
FAILED: src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmlcache_loader.cpp.o 
c++ -Isrc/libZeGrapherLib.a.p -Isrc -I../src -Isubprojects/zecalculator/include -I../subprojects/zecalculator/include -I/usr/include/qt6/QtCore -I/usr/include/qt6 -I/usr/lib64/qt6/mkspecs/linux-g++ -I/usr/include/qt6/QtGui -I/usr/include/qt6/QtWidgets -I/usr/include/qt6/QtNetwork -I/usr/include/qt6/QtSvg -I/usr/include/qt6/QtQuick -I/usr/include/qt6/QtQmlMeta -I/usr/include/qt6/QtQmlWorkerScript -I/usr/include/qt6/QtQmlModels -I/usr/include/qt6/QtOpenGL -I/usr/include/qt6/QtQml -I/usr/include/qt6/QtQmlIntegration -I/usr/include/qt6/QtQuickWidgets -I/usr/include -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -Werror -std=c++20 -O0 -g '-D QT_QML_DEBUG' -pipe -fPIC -DBOOST_ALL_NO_LIB -DQT_QUICKWIDGETS_LIB -DQT_QUICK_LIB -DQT_QMLMETA_LIB -DQT_QMLWORKERSCRIPT_LIB -DQT_QMLMODELS_LIB -DQT_OPENGL_LIB -DQT_QML_LIB -DQT_QMLINTEGRATION_LIB -DQT_SVG_LIB -DQT_NETWORK_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_CORE_LIB -MD -MQ src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmlcache_loader.cpp.o -MF src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmlcache_loader.cpp.o.d -o src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmlcache_loader.cpp.o -c src/zegrapher_highlighter_qmlcache_loader.cpp
cc1plus: fatal error: src/zegrapher_highlighter_qmlcache_loader.cpp: No such file or directory
compilation terminated.
[4/45] Compiling C++ object src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmltyperegistrations.cpp.o
FAILED: src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmltyperegistrations.cpp.o 
c++ -Isrc/libZeGrapherLib.a.p -Isrc -I../src -Isubprojects/zecalculator/include -I../subprojects/zecalculator/include -I/usr/include/qt6/QtCore -I/usr/include/qt6 -I/usr/lib64/qt6/mkspecs/linux-g++ -I/usr/include/qt6/QtGui -I/usr/include/qt6/QtWidgets -I/usr/include/qt6/QtNetwork -I/usr/include/qt6/QtSvg -I/usr/include/qt6/QtQuick -I/usr/include/qt6/QtQmlMeta -I/usr/include/qt6/QtQmlWorkerScript -I/usr/include/qt6/QtQmlModels -I/usr/include/qt6/QtOpenGL -I/usr/include/qt6/QtQml -I/usr/include/qt6/QtQmlIntegration -I/usr/include/qt6/QtQuickWidgets -I/usr/include -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -Werror -std=c++20 -O0 -g '-D QT_QML_DEBUG' -pipe -fPIC -DBOOST_ALL_NO_LIB -DQT_QUICKWIDGETS_LIB -DQT_QUICK_LIB -DQT_QMLMETA_LIB -DQT_QMLWORKERSCRIPT_LIB -DQT_QMLMODELS_LIB -DQT_OPENGL_LIB -DQT_QML_LIB -DQT_QMLINTEGRATION_LIB -DQT_SVG_LIB -DQT_NETWORK_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_CORE_LIB -MD -MQ src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmltyperegistrations.cpp.o -MF src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmltyperegistrations.cpp.o.d -o src/libZeGrapherLib.a.p/meson-generated_.._zegrapher_highlighter_qmltyperegistrations.cpp.o -c src/zegrapher_highlighter_qmltyperegistrations.cpp
src/zegrapher_highlighter_qmltyperegistrations.cpp: In function ‘void qml_register_types_zegrapher_highlighter()’:
src/zegrapher_highlighter_qmltyperegistrations.cpp:23:34: error: ‘Highlighter’ was not declared in this scope
   23 |     qmlRegisterTypesAndRevisions<Highlighter>("zegrapher.highlighter", 1);
      |                                  ^~~~~~~~~~~
src/zegrapher_highlighter_qmltyperegistrations.cpp:23:46: error: no matching function for call to ‘qmlRegisterTypesAndRevisions<<expression error> >(const char [22], int)’
   23 |     qmlRegisterTypesAndRevisions<Highlighter>("zegrapher.highlighter", 1);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from src/zegrapher_highlighter_qmltyperegistrations.cpp:7:
/usr/include/qt6/QtQml/qqml.h:919:6: note: candidate: ‘template<class ... Args> void qmlRegisterTypesAndRevisions(const char*, int, QList<int>*)’
  919 | void qmlRegisterTypesAndRevisions(const char *uri, int versionMajor, QList<int> *qmlTypeIds)
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/qt6/QtQml/qqml.h:919:6: note:   template argument deduction/substitution failed:
src/zegrapher_highlighter_qmltyperegistrations.cpp:23:46: error: template argument 1 is invalid
   23 |     qmlRegisterTypesAndRevisions<Highlighter>("zegrapher.highlighter", 1);
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
</details>

The commit just before, that doesn't use `qml_module`, compiles and runs fine.

I'm probably doing something wrong, but I couldn't find something that could help me out in the documentation and test cases of this MR.

How to reproduce
```
git clone --depth=1 https://github.com/AdelKS/ZeGrapher.git --branch meson-qml
cd ZeGrapher
CXX=g++13 meson setup build
cd build
meson compile
```
needs `>=gcc-13`, on ubuntu 22.04 can be obtained with
```
add-apt-repository -y ppa:ubuntu-toolchain-r/test
apt install -y gcc-13
```

chubinou commented:
I removed the 'moc_source' argument as moc will complain about Qml definition in non-header files, and will likely fail to build (qmltyperegistrar will include the .cpp) 

@eli-schwartz 

> This might only be possible by adding a new frameworks test specifically to handle qml, which can raise a SKIP if qt6 isn't installed. As opposed to the current test which simply tests all currently-installed Qt versions.

I went for this

@AdelKS 

> qmlcachegen complains about wrong input

I  handle the case for qml modules without qml files and added a unit test about this

>  The generated .cpp file that calls qmlRegisterTypesAndRevisions seems to not include the header

Unfortunately this is a qt issue (see [QTBUG-93443](https://bugreports.qt.io/browse/QTBUG-93443)  and [QTBUG-87221](https://bugreports.qt.io/browse/QTBUG-87221)). From what I understand the generated JSON doesn't store the relative header path as they want their installed files to be relocatable, as a result the generated code for qmltyperegistrar doesn't know the path of the headers. This is mentioned in Qt documentation [here](https://doc.qt.io/qt-6/qtqml-cppintegration-definetypes.html)

the solution for this is to explicitly add the header path to the include directories of the target. I added a note  in the documentation about this

> How to reproduce

your branch seems to be missing `zc::eval::Cache` and `zc::eval::ObjectCache` definition

chubinou commented:
> This might only be possible by adding a new frameworks test specifically to handle qml, which can raise a SKIP if qt6 isn't installed.

so now, the CI fails because the tests are skipped. Should I :

* blacklist pretty much all runners (as they don't have qt6)
* add qt6 to the runner images (at least the linux ones)

eli-schwartz commented:
> so now, the CI fails because the tests are skipped. Should I :
> 
> * blacklist pretty much all runners (as they don't have qt6)
> * add qt6 to the runner images (at least the Linux ones)

Probably the latter is better because it provides more coverage, though in at least the case of Bionic I suspect the only option is to blacklist it. ;)

> are the images updated before the CI runs?

No, but the image builder runs the testsuite too. Typically we manually ignore failures for Linux jobs when the corresponding image builder succeeds in the same commit.

AdelKS commented:
Thanks for your input @chubinou 

> your branch seems to be missing zc::eval::Cache and zc::eval::ObjectCache definition

You probably re-used the clone you had before, in which case you need to do a little `meson subprojects update --reset` to get the `ZeCalculator` subproject to the proper version.

> the solution for this is to explicitly add the header path to the include directories of the target

I tried in the branch [meson-qml-2](https://github.com/AdelKS/ZeGrapher/compare/meson-qml-2) to put a `meson.build` file that
1. Is at the same level as the QML C++ header file I want to give to `qml_module()`
2. Give that folder to `include_directories` `kwarg`
    - It does add it to the compile commands as I see an `-Isrc/Utils` in the compile commands
But it still doesn't work somehow.

But, if I move the QML C++ header file to the main include folder (in my case `src`), as I did in the [meson-qml-3](https://github.com/AdelKS/ZeGrapher/compare/meson-qml-3) branch, then it works, not sure what the difference is... :thinking:

chubinou commented:
@AdelKS 

you need to add the include dir where the file from `highlighter_qmlmod` are compiled

```meson
executable('ZeGrapher', 'main.cpp', compiled_ui, compiled_ressources, highlighter_qmlmod,
  link_with: zg_static_lib,
  include_directories: [inc, include_directories('Utils')],
  dependencies: [qt6_dep, boost_dep, zecalculator_dep],
install: true)
```

also, if you build a static library as in you first branch `meson-qml` beware that may need to use `Q_INIT_RESOURCE` to import the static entry points you or you have some linker surprises. I added a test case to illustrate the process

klokik commented:
Moved one of our simpler projects to this implementation of modules - so far so good. There is probably a bug in qmlcachegen on handling paths with symlinks, but this is an edge case of my setup. Thanks for the great work!

klokik commented on docs/markdown/Qt6-module.md:
```diff
@@ -160,6 +164,69 @@ This method takes the following keyword arguments:
+    A list of sources to be transpiled into .moc files for manual inclusion.
+  - `headers` (File | string | custom_target | custom_target index | generator_output)[]:
+     A list of headers to be transpiled into .cpp files
+  - `extra_args` string[]: Extra arguments to pass directly to `qt-moc`
```
You have `moc_extra_args` and don't define this parameter in QmlModuleKwArgs

AdelKS commented:
Thank you very much for your input @chubinou. 

I could transfer my whole project to using `qml_module()` (commit [here](https://github.com/AdelKS/ZeGrapher/commit/ae852c16a296d9b0e7b0eb4a76442db1dae2eacb), I got these warnings but they had no consequence in the resulting exe (as far as I could tell)

```
[42/73] Qml type registration for ZeGrapher
Warning: highlighter.h:8: QSyntaxHighlighter is used as base type but cannot be found.
Warning: information.h:38: QObject is used as base type but cannot be found.
Warning: interactivegraph.h:50: Graph is used as base type but cannot be found.
Warning: mathobject.h:33: QObject is used as base type but cannot be found.
Warning: plotstyle.h:30: QObject is used as base type but cannot be found.
Warning: plotstyle.h:30: uint is used as enum type but cannot be found.
Warning: state.h:34: QObject is used as base type but cannot be found.
Warning: zcbase.h:32: QObject is used as base type but cannot be found.
```

It may be worth adding these two links to the documentation:
- General: https://doc.qt.io/qt-6/qtqml-cppintegration-definetypes.html
- Singletons: https://doc.qt.io/qt-6/qml-singleton.html

Cherry-on-top: any plans with adding support for [qmltc](https://doc.qt.io/qt-6/qtqml-qml-type-compiler.html) ? :star_struck:

chubinou commented:
I updated documentation to reflect @klokik & @AdelKS remarks

@AdelKS 
> Warning: highlighter.h:8: QSyntaxHighlighter is used as base type but cannot be found.

maybe some missing includes/dep  ? moc tend to to be lenient about this. Q_MOC_INCLUDE may help about this

also don't you need public inheritance rather than default (private)

> any plans with adding support for [qmltc](https://doc.qt.io/qt-6/qtqml-qml-type-compiler.html) ? 🤩

not in short term, I'm more interested in providing shader compilation (qsb) support

AdelKS commented:
> maybe some missing includes/dep ? moc tend to to be lenient about this. Q_MOC_INCLUDE may help about this

Thanks! Tried this

`highlighter.h`
```c++
#pragma once

#include <QSyntaxHighlighter>
#include <QQuickTextDocument>
#include "Utils/state.h"

/// @brief backend that highlights parsing errors in math expressions in LineEdit
class Highlighter: public QSyntaxHighlighter
{
  Q_OBJECT
  QML_ELEMENT
  // ...
  Q_MOC_INCLUDE("QSyntaxHighlighter")
```
same warning message, I don't know if this is a `meson` thing or a Qt thing, probably the later, this is the dep object I give to `qml_module()`
```
qt6_dep = dependency('qt6', modules: ['Core', 'Gui', 'Widgets', 'Network', 'Svg', 'Quick', 'Qml', 'QuickWidgets'])
```

> not in short term, I'm more interested in providing shader compilation (qsb) support

That will also be very cool ! I may try to get the `qmltc` support added, I'll open a Draft PR and tag you if so.

klokik commented on mesonbuild/modules/_qt.py:
```diff
@@ -631,3 +743,428 @@ def compile_translations(self, state: 'ModuleState', args: T.Tuple, kwargs: 'Com
+            compile_resource_kwargs: ResourceCompilerKwArgs = {
+                'name': target_name,
+                'sources': qrc_resouces,
+                'extra_args': [],
```
I seem to have a case, when I want to pass `--no-zstd` here. Would it be too hard to add this argument to the qml_module() call?

chubinou commented:
@klokik I added the option

I also renamed the `xxx_extra_args` into `xxx_extra_arguments` in an attempt of uniformity with the other commands (`preprocess` and `compile_translations`)

jpakkane commented:
Cygwin failure is unrelated. Dunno what is up with the image builder there, but the Arch failure at least is understandable. If that test needs to be skipped, then it needs to be whitelisted in the code (I don't remember exactly where, but grepping for the error message should find it).

eli-schwartz commented:
I tried to restart the cygwin test because it is supposed to be fixed now. Got an alternative fascinating error:

https://github.com/mesonbuild/meson/actions/runs/12619860632/job/35164883111 works
> solving: 549 tasks, update: no, use test packages: no
> 
> ` 299 install gcc-core                               12.4.0-3                 `

https://github.com/mesonbuild/meson/actions/runs/12579978087/job/35171502458 fails
> solving: 547 tasks, update: no, use test packages: yes
> 
> ` 305 install gcc-core                               15.0.1+20241201-0.1       `

@jon-turney this seems odd somehow...

By the way the vala error triggered by gcc 15 (really, gcc 14 is sufficient to trigger this) was fixed in https://gitlab.gnome.org/GNOME/vala/-/commit/1383ab9f78f81b4ef56d63b38df702253ac8469b

But again, there is an out of date cygwin package. In this case, vala. Vala has some fixes for its terrible codegen, which has a general habit of breaking badly on the new GCC 14 defaults for "[Modern C](https://wiki.gentoo.org/wiki/Modern_C_porting#How_do_I_reproduce_these_bugs.3F)", but vala in cygwin is orphaned (again) and hasn't been updated since 2018...

jon-turney commented:
> this seems odd somehow...

Indeed. I merged some PRs to the GitHub cygwin-install-action today, which I'm guessing is the cause of suddenly (and incorrectly) pulling in test packages...

I think I've fixed it now.  Sorry for the inconvenience.

> But again, there is an out of date cygwin package.

Yeah.

jpakkane commented:
I cherry picked the CI commit to master. If you remove that and rebase, most of the errors should go away.

chubinou commented:
> If you remove that and rebase, most of the errors should go away.

something doesn't looks right with the fedora/suse docker images

if i fetch the lastest docker image, the qt6 pacakges doesn't appeared as installed 

```
docker run --rm -ti  --mount type=bind,source=$(pwd),target=/work -w /work mesonbuild/fedora:latest /bin/bash

[root@e2ead76ba693 meson]# rpm -qa | grep qt6 | wc -l
0
```

if I install manually the packages (list taken from the install.sh) from there then it works as expected
```
docker run --rm -ti  --mount type=bind,source=$(pwd),target=/work -w /work mesonbuild/fedora:latest /bin/bash

[root@e2ead76ba693 meson]# dnf install qt6-qtdeclarative-devel qt6-qtbase-devel qt6-qttools-devel qt6-linguist qt6-qtbase-private-devel
[root@e2ead76ba693 meson]# rpm -qa | grep qt6 | wc -l
26
[root@e2ead76ba693 meson]# python3 run_project_tests.py -v --only "frameworks/39 qt qml" 
[...]
Total passed tests:  3
Total failed tests:  0
Total skipped tests: 0
```

regarding the unity build failure, the generated code can't be used with unity build, this will generate duplicate symbols, I'll override the unity flag in the tests

chubinou commented:
> something doesn't looks right with the fedora/suse docker images

well, master CI did fail, this explain why the image were not updated

https://github.com/mesonbuild/meson/actions/runs/12642215836/job/35226093290

jpakkane commented:
Master is green, so rebase should fix it?

chubinou commented:
> Master is green, so rebase should fix it?

thanks for the fix, though the opensuse docker image doesn't seems updated yet 

```
docker pull mesonbuild/opensuse:latest
docker run --rm -ti mesonbuild/opensuse:latest /bin/bash          
c31ae6ac0cee:/ # rpm -qa | grep -i qt6 | wc -l
0
c31ae6ac0cee:/ # 
```

> We try to rebuild images weekly, and only deploy them if tests still pass

I guess I need to wait a couple of day, however the tests on fedora are passing now  :thinking:

jpakkane commented on test cases/frameworks/39 qt qml/meson.build:
```diff
@@ -0,0 +1,110 @@
+  version: '1.0',
+  qml_sources: files('Basic.qml', 'subdir/Thing.qml'),
+  qml_singletons: files('QmlSingleton.qml'),
+  qml_internals: files('Internal.qml'),
```
All of these are done with `files()`. We should also support doing this:

```
qml_sources: ['Basic.qml', 'subdir/Thing.qml']
```

so some of the tests should be changed to this form.

chubinou replied:
I  changed a test to use sources as string an one to use generated targets

eli-schwartz commented:
> You can either try to fix those issues, or have this PR assume that the new test will be skipped for lack of qtdeclarative. If/when the errors go away we will update the test to stop being skipped.


> I guess I need to wait a couple of day, however the tests on fedora are passing now 🤔


As I said, "have this PR assume that the new test will be skipped". There is *no* point waiting a couple of days. When I said "We try to rebuild images weekly" that was totally unrelated (we do that in addition to rebuilding them *every* time the image configs change, which you did here, and image failures are NOT a CI failure, so you must verify that your changes build successfully *in this PR* or not depend on them).

Your test assumptions are incompatible with the OpenSUSE image until its unrelated failure is fixed, which means we need to mark it as skipped.

This PR was merged with failing CI @jpakkane
